In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn
from sklearn.linear_model import LinearRegression

In [3]:
data=pd.read_csv("1.02. Multiple linear regression.csv")

In [4]:
data

SAT   GPA  Rand 1,2,3
0   1714  2.40           1
1   1664  2.52           3
2   1760  2.54           3
3   1685  2.74           3
4   1693  2.83           2
..   ...   ...         ...
79  1936  3.71           3
80  1810  3.71           1
81  1987  3.73           3
82  1962  3.76           1
83  2050  3.81           2

[84 rows x 3 columns]

In [5]:
data.head()

SAT   GPA  Rand 1,2,3
0  1714  2.40           1
1  1664  2.52           3
2  1760  2.54           3
3  1685  2.74           3
4  1693  2.83           2

In [8]:
x=data[["SAT","Rand 1,2,3"]]
y=data["GPA"]

In [10]:
x.shape

(84, 2)

In [11]:
reg=LinearRegression()

In [12]:
reg.fit(x,y)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [13]:
reg.coef_

array([ 0.00165354, -0.00826982])

In [14]:
reg.intercept_

0.29603261264909486

## R squared

In [16]:
reg.score(x,y)

0.4066811952814285

## Adjusted R squared Value

$R^2_{adj.}=1-(1-R^2)*\frac{n-1}{n-p-1}$

n=number of observation=84
p=num. of predication=2

In [18]:
x.shape

(84, 2)

In [19]:
n=84
p=2

In [23]:
adjR=1-(1-(reg.score(x,y)))*((n-1)/(n-p-1))

In [24]:
adjR

0.39203134825134023

# F selection
selecting which independent variable is useful and which is not using f variable:

1) Find the p_values and plot the simple reg the ones whoes p_vales>0.05 reject them

In [25]:
from sklearn.feature_selection import f_regression

In [26]:
f_regression(x,y)

(array([56.04804786,  0.17558437]), array([7.19951844e-11, 6.76291372e-01]))

In [27]:
p_values=f_regression(x,y)[1]

In [28]:
p_values

array([7.19951844e-11, 6.76291372e-01])

In [30]:
p_values.round(3)

array([0.   , 0.676])

p_value of SAT=0.000 and Rand 1,2,3=0.676

## Creating Summary tables

In [36]:
reg_summary=pd.DataFrame(data=x.columns.values,columns=["Feature"])

In [38]:
reg_summary

Feature
0         SAT
1  Rand 1,2,3

In [39]:
reg_summary["Coefficent"]=reg.coef_
reg_summary["p values"]=p_values.round(3)

In [40]:
reg_summary

Feature  Coefficent  p values
0         SAT    0.001654     0.000
1  Rand 1,2,3   -0.008270     0.676